crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2289,24 +2297,33 @@ impl<'db> TypeInferenceBuilder<'db> {
                 .map(Type::IntLiteral)
                 .unwrap_or_else(|| builtins_symbol_ty(self.db, "int").to_instance(self.db)),
 
+            // Division by zero
+            // TODO: handle division by zero for floats
+            (
+                Type::IntLiteral(_),
+                Type::IntLiteral(0),
+                ast::Operator::Div | ast::Operator::FloorDiv | ast::Operator::Mod,
+            ) => {
+                self.division_by_zero_diagnostic(binary.into(), left_ty);
+                Type::Unknown
+            }
+
             (Type::IntLiteral(_), Type::IntLiteral(_), ast::Operator::Div) => {
-                // TODO: division by zero error
                 builtins_symbol_ty(self.db, "float").to_instance(self.db)
             }
 
             (Type::IntLiteral(n), Type::IntLiteral(m), ast::Operator::FloorDiv) => n
                 .checked_div(m)
                 .map(Type::IntLiteral)
-                // TODO: division by zero error
-                // It should only be possible to hit this with division by zero, not an overflow.
-                // The overflow case is `i64::MIN // -1` and the negative integer is transformed
-                // into a positive integer prior to here (in the AST).
-                .unwrap_or_else(|| builtins_symbol_ty(self.db, "int").to_instance(self.db)),
+                // It should only impossible to hit this as we handle division by zero above and
+                // overflows are encountered upstream of this inference. The overflow case is
+                // `i64::MIN // -1` and the negative integer is transformed into a positive integer
+                // prior to here (in the AST).
+                .unwrap_or(Type::Unknown),
```

Comment:
I'd be fine with turning this into an `expect` -- if the comment above is wrong, I'd rather find out loudly so we can consider that case, rather than just silently get Unknown.

Reply:
I did some refactoring here in light of the desire for retaining the return types as in https://github.com/astral-sh/ruff/pull/13576#discussion_r1781740580